### DISCUSSION TITLE
Get a single command from a line of multiple semicolon-separated statements

### DISCUSSION THREAD

**Discussion by eugenesvk** [xonsh/xonsh#5085] [Q&A] [answered]
[Fish](https://fishshell.com/docs/current/cmds/commandline.html) has `commandline -p` command that allows you to get only one command out of many (it "Stops at logical operators, terminators, and pipes")

Is there a way for a xonsh xontrib to ask the command line parser to only give you the first `echo '1;2'` in `echo '1;2';echo 2`? Ultimately I'd need the command at the cursor's position and its coordinates so I can replace it with an ouput of anothe tool. Is this possible?

The [PTK](https://python-prompt-toolkit.readthedocs.io/en/master/pages/reference.html) docs don't seem to mention anything, the [pipeliner xontrib](https://github.com/anki-code/xontrib-pipeliner) has a patched version of [asttokens](https://github.com/gristlabs/asttokens), but then it seems to break on simple lack of a space after a semicolon `echo 1;echo 2`

**Comment by anki-code** [comment]
AFAIK the start point could be in [run_subproc](https://github.com/xonsh/xonsh/blob/ec378090ce22abcb10681015579e431ee865f87f/xonsh/procs/specs.py#L869-L887) where command is splitted by parts.

**Reply by eugenesvk** [reply]
Thanks for the suggestion. This function operates on a pre-split list of commands. As far as I understand, I can use `XSH.execer.parser.lexer.split` to split a string I get from `event.current_buffer.document.text`, but then I'd lose information on where the cursor is at, so I wouldn't be able to recreate the original line and might also not even be able to reliably learn of what command the cursor is at?

(I guess, that was the main point of [asttokens](https://github.com/gristlabs/asttokens) - to get tokens connected to source code position)

**Comment by eugenesvk** [comment, marked as answer]
Turns out `from xonsh.lexer get_tokens` is capable enough, so `tokens = get_tokens(text, tolerant=True) # → LexToken(type, value, lineno, lexpos)` (not sure what `tolerant` tolerates here) gets you a generator of tokens that can be used to find find separators of type `sep_type = ['SEMI','PIPE','AND','OR']#; | && ||` and get their `lexpos` position, which then can be used to find what your command is based on cursor position

**Reply by anki-code** [reply]
If you have a snippet of code that demonstrates this please add as well for future researchers. Thanks!
Also PRs into the main branch is welcome!

**Reply by eugenesvk** [reply]
Used it in this [xontrib-navi](https://github.com/eugenesvk/xontrib-navi), here is a function that gives you a sorted list of the positions for any of the 4 command separators  https://github.com/eugenesvk/xontrib-navi/blob/2af4bb41506564a631d9b9c21f32193ff83ca8db/xontrib/navi.py#L62

the list is then used here https://github.com/eugenesvk/xontrib-navi/blob/2af4bb41506564a631d9b9c21f32193ff83ca8db/xontrib/navi.py#LL105C20-L105C31
to find the actual command and everything before/after it
